### PR TITLE
Initial implementation of Agile API

### PR DESCRIPTION
**Pull request by romaninsh**
Wow, this already works pretty well!

### CODE REVIEW

File: examples/test.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        $c->tryLoadBy('name', $this['name']);
+        if ($c->loaded() && $c->id != $this->id) {
+            $errors['name'] = 'Country name must be unique';
+        }
+
+        return $errors;
+    }
+}
+session_start();
+$db = new \atk4\data\Persistence_SQL('mysql:dbname=atk4;host=localhost', 'root', 'root');
```
Comment (DarkSide666) [inline examples/test.php:53]: I don't like this - hard-coded db name and user/password. Should be some kind of config.php file somewhere in root or something like that. Also where is DB build script which create country table?

File: examples/test.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+include '../vendor/autoload.php';
+
+$api = new \atk4\api\Api();
+
+class Country extends \atk4\data\Model
```
Comment (DarkSide666) [inline examples/test.php:7]: Probably test model should be moved somewhere outside test script.

File: src/Api.php [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                        $ret,
+                        200,
+                        [],
+                        JSON_PRETTY_PRINT | JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_AMP | JSON_HEX_QUOT
+                    );
+            }
+
+            if ($this->emitter) {
+                // cannot do anything about it
+                $emitter->emit($this->response);
```
Comment (DarkSide666) [inline src/Api.php]: should be $this->emitter here

File: src/Api.php [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+namespace atk4\api;
+
+/**
+ * Main API class.
+ */
+class Api
+{
+    // @var \Zend\Diactoros\ServerRequest Request object
+    public $request;
```
Comment (DarkSide666) [inline src/Api.php]: More likely these two (request and response) can be set as protected too, but ... maybe they can be used in some way outside of the class too for debug for example.

File: src/Api.php [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            }
+
+            // good until the end
+            if ($r == '**') {
+                break;
+            }
+
+            return false;
+
+            /* never used code
```
Comment (DarkSide666) [inline src/Api.php]: never used code

File: src/Api.php [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+     * @return mixed
+     */
+    public function match($pattern, $callable = null)
+    {
+        $path = explode('/', rtrim($this->path, '/'));
+        $pattern = explode('/', rtrim($pattern, '/'));
+
+        $vars = [];
+
+        // $sanity = 50; unused variable
```
Comment (DarkSide666) [inline src/Api.php]: unused variable

File: src/Api.php [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+                    );
+            }
+
+            if ($this->emitter) {
+                // cannot do anything about it
+                $this->emitter->emit($this->response);
+            }
+
+            // no emitter (is that possible at all ?)
+            return $ret;
```
Comment (DarkSide666) [inline src/Api.php:143]: is it possible at all to not have emitter defined? it's always defined in constructor.
in what case i exclusively would want to set `$api->emitter = false` like in test script?

File: src/Api.php [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+            return $model->load($id)->set($this->requestData)->save()->get();
+        });
+        $this->post($pattern.'/:id', function ($id) use ($model) {
+            return $model->load($id)->set($this->requestData)->save()->get();
+        });
+        $this->delete($pattern.'/:id', function ($id) use ($model) {
+            return !$model->load($id)->delete()->loaded();
+        });
+
+        $this->post($pattern.'/', function () use ($model) {
```
Comment (DarkSide666) [inline src/Api.php]: why we need `'/'` here ?

File: tests/ApiTester.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                'Content-Type'  => 'application/json',
+            ]
+        );
+
+        if ($data !== null) {
+            $request->getBody()->write(json_encode($data));
+        }
+
+        $api = new \atk4\api\Api($request);
+        $api->emitter = false; // don't emmit response
```
Comment (DarkSide666) [inline tests/ApiTester.php:54]: i changed emitter property to protected so this will fail. what's the use-case when we don't want emitter?

**Review by DarkSide666 — CHANGES_REQUESTED** [review]

File: examples/test.php (hunk 1 above)
Comment (romaninsh) [reply examples/test.php:53]: probably will move into memory / sqlite.

File: examples/test.php (hunk 2 above)
Comment (romaninsh) [reply examples/test.php:7]: This was a first quick test. Probably would remove this file long-term or clean it up.

File: src/Api.php (hunk 4 above)
Comment (romaninsh) [reply src/Api.php]: i'm using them publicly though.

File: src/Api.php (hunk 7 above)
Comment (romaninsh) [reply src/Api.php:143]: yes, if we don't want response to be sent out through echo. That's a valid situation when we doing unit tests.

File: tests/ApiTester.php (hunk 9 above)
Comment (romaninsh) [reply tests/ApiTester.php:54]: yep. can you change it back please?

**Comment by romaninsh** [comment]
Comments addressed, proceeding with initial merge

**Pull request merged**